Full CES Reconciliation in the Operator

alan-kut opened the pull request:
Implement full CES reconciliation logic in the Cilium Operator.

Refactor and unify CEPs and CESs watchers in the operator.
Refactor CES Controller:
- Limit the functionality of the Manager to maintain mapping between CEPs and CESs.
- Implement reconciliation in the Reconcilar based on the mapping from the Manager and CEPs and CESs from the watchers stores.
- React to the CESs events in addition to the CEPs events.

This PR should fix all the problems related to CESs out of sync.
If there is a broken state (stale, duplicated, missing CESs) the operator should always fully reconcile it on startup and maintain good state.
If different client breaks CES operator would fix it back.

Fixes: #24581

```release-note
Implement full CES reconciliation logic in the operator
```

dlapcevic commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -8,104 +8,136 @@ import "github.com/cilium/cilium/pkg/lock"
-	defer c.cesMutex.RUnlock()
-	name, ok := c.cepNametoCESName[CEPName(cepName)]
-	return string(name), ok
+// Return CES to which the given CEP is amm
```
typo: CEP is assigned

alan-kut replied:
Fixed

dlapcevic commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -8,104 +8,136 @@ import "github.com/cilium/cilium/pkg/lock"
+func (c *CESToCEPMapping) deleteCES(cesName CESName) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	delete(c.cesNameToCEPNameSet, cesName)
```
Should it also be removed from `c.cesData`?

alan-kut replied:
Yes, fixed.

dlapcevic commented on operator/pkg/ciliumendpointslice/manager_test.go:
```diff
@@ -6,241 +6,108 @@ package ciliumendpointslice
-		m.InsertCEPInCache(cep3, "kube-system")
-		assert.Equal(t, m.getCESCount(), 2, "Total number of CESs allocated is 2")
-		assert.Equal(t, m.getTotalCEPCount(), 3, "Total number of CEPs inserted is 3")
+func TestFCFSManager(t *testing.T) {
```
Consider adding a test with different namespaces, to verify that each CES can only take CEPs from a single namespace.
`UpdateCEPMapping(getCEP("c1", 0), "ns-a")`
`UpdateCEPMapping(getCEP("c2", 0), "ns-b")`

Same for `TestIdentityManager()` below.

alan-kut replied:
Added

dlapcevic commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -53,11 +52,14 @@ const (
 	// Default CES Synctime, multiple consecutive syncs with k8s-apiserver are
 	// batched and synced together after a short delay.
 	DefaultCESSyncTime = 500 * time.Millisecond
+	UpdateEvent        = 1
```
nit: Should it be exported (capital first letter)?
Anyway maybe it's better to make it like this:
```
type EventType int

const (
	AddEvent EventType = iota
	UpdateEvent
	DeleteEvent
)
```

And then update EndpointEvent.event to be EventType.

alan-kut replied:
Done

dlapcevic commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -102,24 +104,47 @@ func GetCEPNameFromCCEP(cep *capi_v2a1.CoreCiliumEndpoint, namespace string) str
+
+	cepWatcher.AddOnUpdateHandler(controller.onEndpointUpdate)
+	cepWatcher.AddOnDeleteHandler(controller.onEndpointDelete)
+	cesWatcher.AddOnDeleteHandler(controller.onSliceUpdate)
```
typo: Should be `AddOnUpdateHandler`

alan-kut replied:
Fixed.

dlapcevic commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -214,396 +99,100 @@ func uniqueCESliceName(desiredCESs *CESToCEPMapping) string {
-func (c *cesMgr) createCES(name string) *cesTracker {
-	var cesName string = name
+func (c *cesMgr) createCES(name, ns string) CESName {
+	var cesName = name
```
nit: `cesName := name`

dlapcevic commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -214,396 +99,100 @@ func uniqueCESliceName(desiredCESs *CESToCEPMapping) string {
 	// If given cep object isn't packed in any of the CES. find a new ces
 	// to pack this cep.
-	cb, cesName := func() (*cesTracker, string) {
+	for {
```
Is the loop needed because `getLargestAvailableCESForNamespace()` can return a `cesName` that can get full before locking the cache, even though the function returns the name of the largest non-full CES?

alan-kut replied:
Yes, that's the reason

alan-kut replied:
Locked sooner so the loop is not needed here.

dlapcevic commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -214,396 +99,100 @@ func uniqueCESliceName(desiredCESs *CESToCEPMapping) string {
+	return ""
 }
 
 // getLargestAvailableCESForNamespace returns the largest CES from cache for the
```
Update the comment that it returns the name of the largest, and that it doesn't guarantee that the CESs is not full by the time it's read from cache by name.

alan-kut replied:
I think it should be clear that if this only returns a name of CES the state may change in the meantime.

dlapcevic commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -612,133 +201,109 @@ func (c *cesManagerIdentity) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint,
+	cesSlice := c.identityToCES[id]
+	for i, ces := range cesSlice {
+		if ces == cesName {
+			cesSlice[i] = cesSlice[len(cesSlice)-1]
```
I don't understand this. How is cesSlice ever reducing the number of elements?

alan-kut replied:
CES is deleted when all the elements inside are deleted.

alan-kut replied:
Fixed the code:
- removed second mutex (everything is now on one mutex)
- added deleting CES from the local operator cache

dlapcevic commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -29,71 +35,174 @@ func newReconciler(client clientset.CiliumV2alpha1Interface, cesMgr operations)
+	// insert missing
+	ceps := r.cesManager.getCEPinCES(cesName)
+	desiredEndpoints := make([]cilium_v2a1.CoreCiliumEndpoint, 0, len(ceps))
+	desiredObjects := make(map[CEPName]*cilium_v2a1.CoreCiliumEndpoint)
```
It's a little difficult to follow `desiredEndpoints` and `desiredObjects`.
Consider renaming `desiredEndpoints` to `updatedEndpoints`, and `desiredObjects` to `desiredEndpoints`.

alan-kut replied:
Simplified a code a little bit. It may reshuffle the CEPs on update now but it's much easier to track

dlapcevic commented on operator/watchers/ciliumendpoint/utils.go:
```diff
@@ -0,0 +1,39 @@
+
+// HasCE returns true or false if the Cilium Endpoint store has the endpoint
+// with the given name.
+func HasCE(ns, name string) (*ciliumApiV2.CiliumEndpoint, bool, error) {
```
The function name and description isn't matching. It should be `GetCE`.

alan-kut replied:
Done

dlapcevic reviewed (approved):
/lgtm

I'll review the changes for "Next step: upgrade the code to follow the same modularization apprach for the CES controller" when it's ready.
Please make it a separate commit.

alan-kut committed:
operator: Implement Full Reconciliation of CES

Refactor CES Controller:
- Limit the functionality of the Manager to maintain mapping between CEPs and CESs.
- Implement reconciliation in the Reconcilar based on the mapping from the Manager and CEPs and CESs from the watchers stores.
- React to the CESs events in addition to the CEPs events.

Signed-off-by: Alan Kutniewski <kutniewski@google.com>

alan-kut committed:
operator/k8s: Add CiliumEndpoint resource

Signed-off-by: Alan Kutniewski <kutniewski@google.com>

alan-kut committed:
operator/k8s: Add CiliumEndpointSlice resource

Signed-off-by: Alan Kutniewski <kutniewski@google.com>

alan-kut committed:
operator: Migrate CES Controller to hive

Signed-off-by: Alan Kutniewski <kutniewski@google.com>

alan-kut committed:
operator: Migrate Identity GC to use CEP and CES from hive.

Signed-off-by: Alan Kutniewski <kutniewski@google.com>

joamaki commented on operator/cmd/root.go:
```diff
@@ -153,6 +153,8 @@ var (
 			// refactored into a proper cell.
 			identitygc.Cell,
+
+			ciliumendpointslice.Cell,
```
Add a short one or two line comment here describing the functionality of this cell. It helps getting an overview of what each top-level cell is doing and how they relate to each other.

alan-kut replied:
Done

joamaki commented on operator/pkg/ciliumendpointslice/cell.go:
```diff
@@ -0,0 +1,57 @@
+	CESWriteQPSBurst = "ces-write-qps-burst"
+)
+
+var Cell = cell.Module(
```
Add descriptive comment here describing what this cell does. E.g. think of the `var Cell` definition as something someone new to the codebase would be hovering over in the `root.go` file and trying to piece together what's going on. Having clear description here about what this implements makes the codebase easier to understand and navigate.

(Incidentally have an issue to "document how to document" hive cells :smile:)

alan-kut replied:
Added

joamaki commented on operator/pkg/ciliumendpointslice/controller.go:
```diff
@@ -0,0 +1,111 @@
+	CiliumEndpointSlice resource.Resource[*v2alpha1.CiliumEndpointSlice]
+
+	Cfg Config
+	//daemonCfg *option.DaemonConfig
```
Drop this?

alan-kut replied:
Fixed. I wasn't sure how to propagate option.Config.EnableCiliumEndpointSlice to CES Controller but at the end I used it directly so at the end it's not needed and I forgot to remove it.

joamaki commented on operator/pkg/ciliumendpointslice/controller.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	cesController := &Controller{
+		logger:                         p.Logger.WithField(logfields.LogSubsys, "ces-controller"),
```
The `p.Logger` already has `LogSubsys` set to `k8s-ces-controller` via `cell.Module`, so no need to set it again.

alan-kut replied:
Fixed

joamaki commented on operator/pkg/ciliumendpointslice/controller.go:
```diff
@@ -0,0 +1,111 @@
+
+	p.Lifecycle.Append(hive.Hook{
+		OnStart: func(context hive.HookContext) error {
+			return cesController.Start(context)
```
It would be preferable to do `p.Lifecycle.Append(cesController)` as then the `cilium-agent hive` output that shows the start and stop hooks would have nice function names rather than anonymous functions.

`Append` takes a `HookInterface` containing the `Start` and `Stop` methods. `Hook` is an alternative way of implementing the hooks, but it's usually preferably to have an object implement the `HookInterface` directly.

alan-kut replied:
Fixed, thanks. It makes sense.

joamaki commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -7,93 +7,209 @@ import (
-func newReconciler(client clientset.CiliumV2alpha1Interface, cesMgr operations) *reconciler {
+func newReconciler(client clientset.CiliumV2alpha1Interface, cesMgr operations,
+	logger logrus.FieldLogger, ciliumEndpoint resource.Resource[*cilium_v2.CiliumEndpoint], ciliumEndpointSlice resource.Resource[*cilium_v2a1.CiliumEndpointSlice]) *reconciler {
+	cepStore, _ := ciliumEndpoint.Store(context.Background())
```
Do not call blocking (Store blocks until resource has synchronized) or side-effectful methods from a constructor as this is also called when inspecting hive (e.g. `go run ./operator hive`). Instead call these from a `Start` hook and use the start context to allow aborting the start.

https://docs.cilium.io/en/latest/contributing/development/hive/#guidelines

alan-kut replied:
This is called on Start - CES Controller creates reconciler (using newReconciler).
I moved this to Start just because of this.

I see that newReconciler may be confusing. I don't have much better idea as it doesn't do anything else except for getting those stores, logger and client.

joamaki replied:
Ah sorry, I missed that it was called from a start hook. In that case I don't have any problem with this. I think the naming etc. is fine here. I should've checked how it was used and not assume.

joamaki commented on operator/watchers/ciliumendpoint/cilium_endpoint.go:
```diff
@@ -0,0 +1,153 @@
+type cepHanlderFunc func(cep *ciliumApiV2.CiliumEndpoint)
+
+var (
+	log      = logging.DefaultLogger.WithField(logfields.LogSubsys, "watchers")
```
s/watchers/ciliumendpoint/

alan-kut replied:
Done

joamaki commented on operator/watchers/ciliumendpoint/cilium_endpoint.go:
```diff
@@ -0,0 +1,153 @@
+func CiliumEndpointsInit(ctx context.Context, wg *sync.WaitGroup, clientset k8sClient.Clientset) {
+	once.Do(func() {
+		CiliumEndpointStore = cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc, indexers)
+		ciliumEndpointInformer := informer.NewInformerWithStore(
```
We should not have two informers for the same object kind. Is this still needed after adding `Resource[*CiliumEndpoint]`? Would it be hard to convert the remaining uses? If it is consider converting this internally to use `Resource[*CiliumEndpoint]` as an intermediate step towards removing it. https://github.com/cilium/cilium/blob/main/pkg/k8s/watchers/pod.go#L124

alan-kut replied:
When I started working on the Full CES Reconciliation I was unaware of the hive.
I started by fixing the watchers, implemented full reconciliation and then I learned about the hive and I moved CES Controller to hive.

If I knew about the hive I wouldn't refactor legacy watchers. I can revert this refactoring if needed.
I'll try to revert this refactoring.
This won't help with the main issue - there are still use cases on legacy watchers.

I talked to @pippolo84 about this problem and it's tricky one. It may be hard to move everything out of legacy watchers at once. He agreed that we may have some period of time in which we would have two informers.

I'm trying to move other use case to hive in the #27458
I could work on all remaining use cases but there is already some work done by @pippolo84 so I'm not sure whether it's ok to hijack it.

joamaki commented on operator/watchers/ciliumendpointslice/cilium_endpoint_slice.go:
```diff
@@ -0,0 +1,119 @@
+func CiliumEndpointSliceInit(ctx context.Context, wg *sync.WaitGroup, clientset k8sClient.Clientset) {
+	once.Do(func() {
+		var ciliumEndpointSliceInformer cache.Controller
+		CiliumEndpointSliceStore, ciliumEndpointSliceInformer = informer.NewInformer(
```
Same here, we really don't want duplicate informers.

joamaki reviewed (changes requested):
Hey, this is looking great!

Any feedback on `pkg/hive` and `pkg/k8s/resource`? Did you find anything confusing or documentation lacking?

pippolo84 commented on operator/identitygc/crd_gc_test.go:
```diff
@@ -0,0 +1,88 @@
+	hive := hive.New(
+		k8sClient.FakeClientCell,
+		k8s.ResourcesCell,
+		cell.Invoke(func(c *k8sClient.FakeClientset, cep resource.Resource[*cilium_v2.CiliumEndpoint], ces resource.Resource[*cilium_v2a1.CiliumEndpointSlice]) error {
```
`cep` parameter is unused, it can be removed

alan-kut replied:
Fixed

pippolo84 commented on operator/identitygc/crd_gc_test.go:
```diff
@@ -0,0 +1,88 @@
+		}),
+	)
+	hive.Start(context.Background())
+	<-wait
```
`Invoke` calls are executed synchronously before `Start` returns, so there's no need to wait here (but I suggest to add a check for the error returned by `hive.Start`).

alan-kut replied:
Fixed. I think I had problems with it at some point but it works now. Removed

pippolo84 commented on operator/identitygc/crd_gc_test.go:
```diff
@@ -0,0 +1,88 @@
+	gotIdentities = usedIdentitiesInCESs(cesStore)
+	assertEqualIDs(t, wantIdentities, gotIdentities)
+
+	hive.Stop(context.Background())
```
Same here, better to check for the error returned by `hive.Stop`

alan-kut replied:
Fixed

pippolo84 commented on operator/identitygc/crd_gc_test.go:
```diff
@@ -0,0 +1,88 @@
+	// 5 IDs in the store.
+	cesA := createCESWithIDs("cesA", []int64{1, 2, 3, 4, 5})
+	fakeClient.CiliumV2alpha1().CiliumEndpointSlices().Create(context.Background(), cesA, meta_v1.CreateOptions{})
+	time.Sleep(50 * time.Millisecond)
```
One-shot `time.Sleep`s with fixed duration don't play nice with the unpredictability of the CI environment. To avoid introducing flakiness, I suggest to use `testutils.WaitUntil` or `testutils.WaitUntilWithSleep` with a very large timeout.

alan-kut replied:
Glad I could learn something new. Fixed.

pippolo84 commented on operator/identitygc/crd_gc_test.go:
```diff
@@ -0,0 +1,88 @@
+	// 10 IDs in the store.
+	cesB := createCESWithIDs("cesB", []int64{10, 20, 30, 40, 50})
+	fakeClient.CiliumV2alpha1().CiliumEndpointSlices().Create(context.Background(), cesB, meta_v1.CreateOptions{})
+	time.Sleep(50 * time.Millisecond)
```
Same here, use `testutils.WaitUntil` or `testutils.WaitUntilWithSleep`

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -8,104 +8,137 @@ import "github.com/cilium/cilium/pkg/lock"
+func (c *CESToCEPMapping) getAllCESs() []CESName {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	var cess []CESName
```
Nit: `cess := make([]CESName, 0, len(c.cesNameToCEPNameSet))`

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -8,104 +8,137 @@ import "github.com/cilium/cilium/pkg/lock"
+func (c *CESToCEPMapping) getCEPsInCES(ces CESName) []CEPName {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	ceps := []CEPName{}
```
Nit: `ceps := make([]CEPName, 0, len(c.cesNameToCEPNameSet[ces]))`

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/controller.go:
```diff
@@ -0,0 +1,97 @@
+
+// registerController creates and initializes the CES controller
+func registerController(p params) {
+	if !p.Clientset.IsEnabled() && option.Config.EnableCiliumEndpointSlice {
```
Shouldn't this be `if !p.Clientset.IsEnabled() || !option.Config.EnableCiliumEndpointSlice {` ?

alan-kut replied:
You are right. Fixed. Moved the EnableCiliumEndpointSlice to shared config and added test to verify it is used correctly.

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice_test.go:
```diff
@@ -51,115 +52,137 @@ func createStoreEndpointSlice(name string, namespace string, endpoints []capi_v2
+		}),
+	)
+	hive.Start(context.Background())
+	<-wait
```
No need to wait here.

alan-kut replied:
Fixed here and in the other places in the CES controller tests

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice_test.go:
```diff
@@ -51,115 +52,137 @@ func createStoreEndpointSlice(name string, namespace string, endpoints []capi_v2
+		}),
+	)
+	hive.Start(context.Background())
+	<-wait
```
No need to wait here.

alan-kut replied:
fixed

pippolo84 reviewed (changes requested):
Thanks for the PR and for the additional effort to refactor the identitygc.
I did a first pass and left some suggestions inline, I'm planning to do another pass focusing on the CES controller inner workings tomorrow.

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -48,263 +48,237 @@ const (
+	} else {
+		c.manager = newCESManagerFcfs(c.maxCEPsInCES, c.logger)
+	}
+	c.reconciler = newReconciler(c.clientset.CiliumV2alpha1(), c.manager, c.logger, c.ciliumEndpoint, c.ciliumEndpointSlice)
```
We should pass down the `hive.HookContext` here in order to use it when Creating/Updating/Deleting CESs in place of `context.TODO()`.

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -7,93 +7,214 @@ import (
-		log.WithError(err).WithFields(logrus.Fields{
-			logfields.CESName: cesToCreate,
+	if _, err = r.client.CiliumEndpointSlices().Create(
+		context.TODO(), newCES, meta_v1.CreateOptions{}); err != nil {
```
Following the above comment, we should use the context from the upper level and extend the check as `err != nil && !errors.Is(err, context.Canceled)`.
Same for the update and deletion.

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice_test.go:
```diff
@@ -54,240 +52,137 @@ func createStoreEndpointSlice(name string, namespace string, endpoints []capi_v2
+	cesN, _ = mapping.getCESName(NewCEPName("cep6", "ns"))
+	assert.Equal(t, cesN, NewCESName("ces2"))
+	cesN, _ = mapping.getCESName(NewCEPName("cep7", "ns"))
+	assert.Equal(t, cesN, NewCESName("ces2"))
```
The call to hive.Stop is missing, this test will leak goroutines.

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice_test.go:
```diff
@@ -54,240 +52,137 @@ func createStoreEndpointSlice(name string, namespace string, endpoints []capi_v2
+
+	cesController.processEnqueuedPreInitEndpoints()
+	time.Sleep(DefaultCESSyncTime * 2) // elements are added to the queue with a default delay
+	assert.Equal(t, 2, cesController.queue.Len())
```
Missing call to hive.Stop

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/reconciler_test.go:
```diff
@@ -4,217 +4,226 @@
+	assert.Equal(t, "ns", createdSlice.Namespace)
+	eps := []string{createdSlice.Endpoints[0].Name, createdSlice.Endpoints[1].Name}
+	assert.Contains(t, eps, "cep1")
+	assert.Contains(t, eps, "cep2")
```
Missing call to hive.Stop

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/reconciler_test.go:
```diff
@@ -4,217 +4,226 @@
+	assert.Equal(t, "ns", updatedSlice.Namespace)
+	eps := []string{updatedSlice.Endpoints[0].Name, updatedSlice.Endpoints[1].Name}
+	assert.Contains(t, eps, "cep1")
+	assert.Contains(t, eps, "cep2")
```
Missing call to hive.Stop

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/reconciler_test.go:
```diff
@@ -4,217 +4,226 @@
+	// ces1 contains cep1 and cep3, but it's mapped to nothing so it should be deleted
+	r.reconcileCES(NewCESName("ces1"))
+
+	assert.Equal(t, "ces1", deletedSlice)
```
Missing call to hive.Stop

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/reconciler_test.go:
```diff
@@ -4,217 +4,226 @@
+	// ces1 contains cep1 and cep3, but it's mapped to nothing so it should be deleted
+	r.reconcileCES(NewCESName("ces1"))
+
+	assert.Equal(t, true, noRequest)
```
Missing call to hive.Stop

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/controller.go:
```diff
@@ -0,0 +1,97 @@
+	Cfg Config
+}
+
+type EndpointEvent struct {
```
No need to export this type.

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -48,263 +48,237 @@ const (
+func (c *Controller) processEnqueuedPreInitEndpoints() {
+	for _, e := range c.preInitEnqueuedEndpointsEvents {
+		if e.event == updateEvent {
+			c.onEndpointUpdate(e.cep)
```
I suggest to add a Debug log here too, as already done in `runCiliumEndpointsUpdater`, it might be convenient. Same for the deletion.

alan-kut replied:
Added

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice_test.go:
```diff
@@ -54,240 +52,131 @@ func createStoreEndpointSlice(name string, namespace string, endpoints []capi_v2
+	cesN, _ = mapping.getCESName(NewCEPName("cep6", "ns"))
+	assert.Equal(t, cesN, NewCESName("ces2"))
+	cesN, _ = mapping.getCESName(NewCEPName("cep7", "ns"))
+	assert.Equal(t, cesN, NewCESName("ces2"))
```
Also, before stopping the hive, you need to call `cesController.queue.ShutDown()` to stop the internal client-go goroutine managing the delaying queue. That is because you are manually managing (outside of the hive lifecycle) the CES controller (and so controller `Stop` is not called).
After doing that, you can get rid of the goleak.IgnoreCurrent, since there will be no other goroutine leak (ultimately making the test better in finding possible leaks).

alan-kut replied:
Fixed stopping the queue.

Regarding IgnoreCurrent - Shouldn't we always use it?
Otherwise it's easy to add a test for A (e.g. some CES controller test leaking) and change some other part B (e.g. Identity GC) that does have such test and the test for B would start failing even though the problem is totally outside of B.

pippolo84 replied:
That's true but, if we keep that, it's easier to add a new test that might leak a goroutine, and let the leak go unnoticed. I find this to be even worse, although you're right that having a failure in another test is not ideal.
The hive based tests are like integration tests for multiple cells even if they are written as unit. I suggest to first add goleak to each hive based test (like the ones in `endpointslice_test.go` and in `reconciler_test.go`) and then check that we are not leaking anything (except for anything that is due to a dependency like client-go). If so, I still suggest to avoid `IgnoreCurrent` for now.

alan-kut replied:
Also - those test would fail if some other test fail and doesn't clean up everything.
```
hive.Start()

err := checkSth()
if err != nil {
  t.Fatalf("failed")
}

hive.Stop()
```

Removed IgnoreCurrent for now.

It would be good to have some kind of a leak test at the end if all other test passed, maybe?

pippolo84 replied:
> Also - those test would fail if some other test fail and doesn't clean up everything.

That's true but that is less harmful, since it'll be clear that the root cause is the first test failure.

> It would be good to have some kind of a leak test at the end if all other test passed, maybe?

I think that right now is not necessary, since we can check every single hive-based test with goleak, excluding what we already know is leaking (and out of our control), like `k8s.io/client-go/util/workqueue.(*Type).updateUnfinishedWorkLoop`.

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice_test.go:
```diff
@@ -54,240 +52,131 @@ func createStoreEndpointSlice(name string, namespace string, endpoints []capi_v2
+
+	cesController.processEnqueuedPreInitEndpoints()
+	time.Sleep(DefaultCESSyncTime * 2) // elements are added to the queue with a default delay
+	assert.Equal(t, 2, cesController.queue.Len())
```
Same here, we need to call `cesController.queue.ShutDown()`before stopping the hive.

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/utils.go:
```diff
@@ -0,0 +1,31 @@
+		if !mapping.hasCESName(NewCESName(cesName)) {
+			return cesName
+		}
+	}
```
Not sure if this truly qualifies as a "hot path", but I suggest to avoid the usage of `fmt.Sprintf` in favor of `strings.Builder` (it is also possible to preallocate the underlying buffer with the right size):
```
	var sb strings.Builder
	for {
		rn1, rn2 := randomName(9), randomName(5)

		sb.Reset()
		sb.Grow(len(cesNamePrefix) + 1 + len(rn1) + 1 + len(rn2))
		sb.WriteString(cesNamePrefix)
		sb.WriteRune('-')
		sb.WriteString(rn1)
		sb.WriteRune('-')
		sb.WriteString(rn2)
		cesName := sb.String()

		if !mapping.hasCESName(NewCESName(cesName)) {
			return cesName
		}
	}
```

See https://github.com/cilium/cilium/issues/19571

alan-kut replied:
I just moved this code but there is no harm in changing it now do string builder.

pippolo84 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -211,394 +97,99 @@ func uniqueCESliceName(desiredCESs *CESToCEPMapping) string {
+			defer c.mutex.Unlock()
+			if c.mapping.countCEPsInCES(cesName) == 0 {
+				c.mapping.deleteCES(cesName)
+			}
```
I suggest to add a brief comment to explain that we're doing the first check with the read lock to avoid the performance penalty of acquiring the write lock every time. The double check might be confusing for a casual reader.

alan-kut replied:
Added, you are right, it may  be confusing

pippolo84 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -607,133 +198,125 @@ func (c *cesManagerIdentity) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint,
+			cesSlice[i] = cesSlice[len(cesSlice)-1]
+		}
 	}
+	c.identityToCES[id] = cesSlice
```
Shouldn't we keep track of how many elements we've overwritten in cesSlice at line 291 and then assign here `cesSlice[:len(cesSlice)-removed]`? Otherwise the slices in the `c.identityToCES` won't never shrink.

alan-kut replied:
Fixed. Good catch.

pippolo84 reviewed (changes requested):
I've completed a second round of review.
Overall LGTM, my suggestions are mainly related to the tests. Thanks again!

pippolo84 commented on operator/pkg/ciliumendpointslice/controller_test.go:
```diff
@@ -0,0 +1,135 @@
+	if err != nil {
+		t.Fatalf("Couldn't verify if CES is created: %s", err)
+	}
+	// Verify CES is created when CES features is disabled
```
Typo: `disabled` -> `enabled`

alan-kut replied:
Fixed

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -48,263 +48,244 @@ const (
+		}).Debug("Enqueueing CES (if not empty name)")
+		if ces.Name != "" {
+			if c.enqueuedAt[ces].IsZero() {
+				c.enqueuedAt[ces] = time.Now()
```
I think `enqueuedAt` needs to be protected from concurrent writes. It is written here (from the workerpool `cilium-endpoint-slices-updater` goroutine) and from the `worker` goroutine in `getAndResetCESProcessingDelay`.

alan-kut replied:
```go
func enqueueCES(ces CESName) {
  if c.enqueuedAt[ces].IsZero() {
    c.enqueuedAt[ces] = time.Now()
  }
}

func getAndReset(ces CESName) {
  enqueued, exists := c.enqueuedAt[ces]
  if exists && !enqueued.IsZero() {
    c.enqueuedAt[ces] = time.Time{}
    return time.Since(enqueued).Seconds()
  }
}
```

So when we enqueue a CES and there is no timestamp we set it to current time. When we process the CES and there is a timestamp we reset it and report time.

The conflict and race may happen if we try to enqueue and process a CES in the same time but the result doesn't matter.

Case I: enqueueAt[ces] was not zero before the race.
In this case the result reported from getAndReset will be correct delay.
At the end enqueueAt[ces] may be zero (if getAndReset finished later than enqueue) or time.Now() otherwise. The same can happen if either enqueue happens fully before or after getAndReset so synchronization doesn't change anything.

Case II: enqueueAt[ces] was zero.
In this case race may happen that enqueue first sees it's zero, changes it to time.now() and then getAndReset sees it's not zero, reports the delay and reset it. The same can happen if enqueueAt happens fully before getAndReset.

I don't see a case in which the result without synchronization cannot be achieved with synchronization. I'd rather not add a lock here just to protect queue delay metric.

That said this may be wrong. The code should always show the time really spent in the queue. I'm not sure how to fix it. It would be great to add new field to the queue. However this would break the behavior of not processing the same items multiple times as all entries would be unique now.
We can't also just prepare a queue of enqueued at as not all add will add element to the queue.

I will keep thinking about it.

pippolo84 replied:
> I don't see a case in which the result without synchronization cannot be achieved with synchronization. I'd rather not add a lock here just to protect queue delay metric.

The write operations themselves (`c.enqueuedAt[ces] = time.Now()` and `c.enqueuedAt[ces] = time.Time{}`) are not atomic. So we can end up with a value in `c.enqueuedAt[ces]` that is neither `time.Now()` nor `time.Time{}`.
That's because `time.Time` is defined as a struct:

```
type Time struct {
	wall uint64
	ext  int64

	loc *Location
}
```

And if that happens, `getAndReset` might return a totally bogus time interval when doing `time.Since(enqueued).Seconds()` the next time is called for the same CES.

alan-kut replied:
Added a lock

rolinh reviewed (changes requested):
I just did a review of the vendor changes. Please, see suggestion below.

rolinh commented on operator/pkg/ciliumendpointslice/controller_test.go:
```diff
@@ -0,0 +1,135 @@
+	if err := testutils.WaitUntil(func() bool {
+		return len(cepStore.List()) > 0
+	}, time.Second); err != nil {
+		return false, errors.Errorf("failed to get CEP: %s", err)
```
I don't see a good argument for having `github.com/pkg/errors` as a new direct dependency. 
```suggestion
		return false, fmt.Errorf("failed to get CEP: %s", err)
```

alan-kut replied:
Fixed, thanks

pippolo84 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -23,75 +18,31 @@ var (
-	// multiple times concurrently, and if CES is added multiple times before it
-	// can be processed, this will only be processed only once.
-	queue workqueue.RateLimitingInterface
+	mutex lock.Mutex
```
This lock can be avoided since `mapping` is assigned once at the manager startup and then every ops on it is protected by its internal lock. Same for `maxCEPsInCES`, it is assigned at startup and then only read.

alan-kut replied:
It is used in the following cases:

Get largest not full CES and create new one if there is none. Without lock there may be too many CEPs in CES or too many CESs (all gets the same CES and fill it above capacity or all create new one when one would be enough)
```go
cesName = c.getLargestAvailableCESForNamespace(ns)
if cesName.Name == "" {
	cesName = c.createCES("", ns)
}
c.mapping.insertCEP(cepName, cesName)
```

Check if CES can be deleted. Without this check new CEP may be added to the CES and CES then deleted.
```go
c.mutex.Lock()
defer c.mutex.Unlock()
if c.mapping.countCEPsInCES(cesName) == 0 {
	c.mapping.deleteCES(cesName)
}
```

The same, a little bit more complicated, for identity.

I think I could rewrite this code without this lock. The mapping could do the checks and then return error. But it means the code would need to be changed to the loop.

pippolo84 replied:
> It is used in the following cases:
> 
> Get largest not full CES and create new one if there is none. Without lock there may be too many CEPs in CES or too many CESs (all gets the same CES and fill it above capacity or all create new one when one would be enough)
> 

But the only operations that can insert or remove a CEP are:
- `UpdateCEPMapping` and `RemoveCEPMapping`, but these are serialized by the channel returned by `Events` in `runCiliumEndpointsUpdater` (IOW it's not possible to handle multiple CiliumEndpoint related events at the same time)
- `initializeMappingCEPtoCES` that is called only before initial sync is completed

So, it should not be possible to have multiple goroutines concurrently reading the CEPs in CES counter and trying to  insert or remove a CEP in a CES. Or am I missing something?

alan-kut replied:
Yes. As long as we are sure we cannot process multiple events from the CEP at the same time we are good here. I'm removing those locks for now.

I thought that the CES controller worker also changes mapping but it's not the case, it's:

CEP event => update mapping => enqueue CES => worker processing CESs from the queue
CES event => enqueue CES => the same worker as above processing CESs from the queu

I'll remove the locks.

alan-kut replied:
Removed

pippolo84 commented on operator/pkg/ciliumendpointslice/controller.go:
```diff
@@ -0,0 +1,100 @@
+
+	enqueuedAt                     map[CESName]time.Time
+	preInitEnqueuedEndpointsEvents []endpointEvent
+	endpointsMappingInitialized    bool
```
I think the access to `endpointsMappingInitialized` needs to be synchronized (for example using an `atomic.Bool`). That's because it is written by the goroutine that executes `syncCESsInLocalCache` and it is read by the goroutine that executes `onEndpointUpdate` and `onEndpointDelete` (that is, the goroutine `cilium-endpoints-updater` from the workerpool).
IOW, the second goroutine is doing a busy wait on that boolean, and doing that without synchronization is incorrect, as reported [here](https://go.dev/ref/mem) (look for the relevant example regarding "busy waiting").

alan-kut replied:
I don't think it's busy waiting.

```go
func (c *Controller) onEndpointEvent(cep *cilium_api_v2.CiliumEndpoint) {
  if c.endpointsMappingInitialized {
    process
  } else {
    add to queue
}
```
Nothing else read it. It's changed to true later.

I agree we need synchronization to avoid situation in which we miss an event - check for c.endpointsMappingInitialized return false but the queue is started being processed before new element is added.

I think we need synchronization on the preInitEnqueuedEndpointsEvents even more. If two threads change it at the same time (it's not possible right now I think as we have 1 thread processing the events) then we will miss some events.

It's too bad we will have quite complicated code here just for short period of time.
I'll try to remove it all together. I think we can start processing the events only after the initial mapping is initialized.

alan-kut replied:
Done it. It is quite simple with the hive as we can start processing event when we want to.

pippolo84 replied:
> I'll try to remove it all together. I think we can start processing the events only after the initial mapping is initialized.

Yep, this sounds better from a readability perspective too. :+1:

pippolo84 reviewed (changes requested):
Thanks for addressing all the comments! I've focused this last round of review on the usage of locks. I think there is something that needs to be addressed.
Apart from that, I think now the PR is starting to looking great! :+1:

rolinh reviewed (approved):
No more vendor changes so the PR lgtm in that respect. Thanks for your contribution!

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -48,263 +41,223 @@ const (
+	if !exists {
+		return 0
+	}
+	c.enqueuedAtLock.Lock()
```
I think we should take the lock before checking the key existence, too. That's because it is possible to have a concurrent write to the `c.enqueuedAt` map at line 109 and that can potentially reallocate the map internal buckets (e.g: think about when the map grows) giving us incorrect results.

alan-kut replied:
Done

pippolo84 reviewed (approved):
Awesome, thanks! :rocket:

asauber reviewed (approved):
Config-related changes LGTM. 

A great reference for Hive considerations as well.
